packages/firestore/src/api/database.ts

```diff
@@ -753,7 +809,10 @@ export class DocumentReference<T = PublicDocumentData>
     return new DocumentReference<U>(
       this.firestore,
       this._delegate.withConverter(
-        converter as UntypedFirestoreDataConverter<U>
+        UntypedFirestoreDataConverterAdapter.getInstance(
+          this.firestore,
+          converter
```

Comment:
If we store this as an instance member (i.e. rather than storing the original converter we store the `UntypedFirestoreDataConverterAdapter`), do we still need the `getInstance` method and its map? The converter would then be the same across calls. You can even pass the UntypedFirestoreDataConverterAdapter down from the CollectionReference to the DocumentReference if you make sure that you never wrap a converter twice.

Reply:
I also realized that we need to key the WeakMap on both Firestore in addition to the converter instance, so that the object specified to the wrapper converter will be associated with the correct Firestore instance.